### PR TITLE
add bitcoin npm lib examples

### PR DESCRIPTION
**Pull request by bdemann**

### CODE REVIEW

File: examples/bitcoinjs-lib/test/tests.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                    const response = await fetch(
+                        `${origin}/sign-bitcoin-message`,
+                        { method: 'POST' }
+                    );
+                    const responseText = await response.text();
+
+                    return {
+                        Ok:
+                            responseText ===
+                            '076595661fbe8eee656defdcb189bdcebeaa34f8dc8115f78c11dfb6f628744b51d16b5d91ff70217084678eeadf479a045eb6530af8d042a7f521622d9a4c38'
```
Comment (lastmjs) [inline examples/bitcoinjs-lib/test/tests.ts:123]: Make them all the same

File: examples/bitcore-lib/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    });
+
+    app.post('/verify-bitcoin-message', (req, res) => {
+        const privateKey = new PrivateKey(
+            'b221d9dbb083a7f33428d7c2a3c3198ae925614d70210e28716ccaa7cd4ddb79'
+        );
+        const message = new Message('This is an example of a signed message.');
+        const signature = message.sign(privateKey);
+        const address = privateKey.toAddress();
+        console.log(signature);
```
Comment (lastmjs) [inline examples/bitcore-lib/src/index.ts:72]: Remove

File: examples/bitcore-lib/test/pretest.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { execSync } from 'child_process';
+
+async function pretest() {
+    await new Promise((resolve) => setTimeout(resolve, 5000));
```
Comment (lastmjs) [inline examples/bitcore-lib/test/pretest.ts:4]: Remove

File: examples/bitcore-lib/test/tests.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+dns.setDefaultResultOrder('ipv4first');
+
+import { Test } from 'azle/test';
+
+export function getTests(canisterId: string): Test[] {
+    const origin = `http://${canisterId}.localhost:8000`;
+
+    return [
+        {
+            name: '/get-address',
```
Comment (lastmjs) [inline examples/bitcore-lib/test/tests.ts:11]: These should match the bitcorejs-lib

File: examples/bitcore-lib/test/pretest.ts (hunk 3 above)
Comment (bdemann) [reply examples/bitcore-lib/test/pretest.ts]: done

File: examples/bitcore-lib/src/index.ts (hunk 2 above)
Comment (bdemann) [reply examples/bitcore-lib/src/index.ts]: done

File: examples/bitcore-lib/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    app.get('/create-transaction', (req, res) => {
+        const privateKey = new PrivateKey(
+            'b221d9dbb083a7f33428d7c2a3c3198ae925614d70210e28716ccaa7cd4ddb79'
+        );
+        const utxo = Transaction.UnspentOutput.fromObject({
+            txId: '115e8f72f39fad874cfab0deed11a80f24f967a84079fb56ddf53ea02e308986',
+            outputIndex: 0,
+            address: privateKey.toAddress(),
+            script: '76a91447862fe165e6121af80d5dde1ecb478ed170565b88ac',
+            satoshis: 50000
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/bitcore-lib/src/index.ts]: done

File: examples/bitcoinjs-lib/test/tests.ts (hunk 1 above)
Comment (bdemann) [reply examples/bitcoinjs-lib/test/tests.ts]: done

File: examples/bitcoinjs-lib/test/tests.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                try {
+                    const response = await fetch(
+                        `${origin}/create-transaction`,
+                        { method: 'POST' }
+                    );
+                    const transactionHex = await response.text();
+
+                    return {
+                        Ok:
+                            transactionHex ===
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/bitcoinjs-lib/test/tests.ts]: done

File: examples/bitcoinjs-lib/test/tests.ts [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        {
+            name: '/get-public-key',
+            test: async () => {
+                try {
+                    const response = await fetch(`${origin}/get-public-key`);
+                    const publicKeyHex = await response.text();
+
+                    return {
+                        Ok:
+                            publicKeyHex ===
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/bitcoinjs-lib/test/tests.ts]: done

File: examples/bitcoinjs-lib/test/tests.ts [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+            name: '/get-address',
+            test: async () => {
+                try {
+                    const response = await fetch(`${origin}/get-address`);
+                    const responseText = await response.text();
+
+                    return {
+                        Ok:
+                            responseText ===
+                            '1PmamxRspvjCV7vDqMpzvKf92epy1utZVj'
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/bitcoinjs-lib/test/tests.ts]: done

File: examples/bitcoinjs-lib/test/pretest.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import { execSync } from 'child_process';
+
+async function pretest() {
+    await new Promise((resolve) => setTimeout(resolve, 5000));
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/bitcoinjs-lib/test/pretest.ts]: done

File: examples/bitcoinjs-lib/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        const validator = (
+            pubkey: Buffer,
+            msghash: Buffer,
+            signature: Buffer
+        ): boolean => ECPair.fromPublicKey(pubkey).verify(msghash, signature);
+        const keyPair = ECPair.fromWIF(
+            'L2uPYXe17xSTqbCjZvL2DsyXPCbXspvcu5mHLDYUgzdUbZGSKrSr'
+        );
+        const psbt = new bitcoin.Psbt();
+        console.log('about to add input');
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/bitcoinjs-lib/src/index.ts]: done

File: examples/bitcoinjs-lib/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+    app.get('/get-private-key-wif', (req, res) => {
+        const keyPair = ECPair.fromWIF(
+            'L3BybjkmnMdXE6iNEaeZTjVMTHA4TvpYbQozc264Lto9yVDis2nv'
+        );
+        res.send(keyPair.toWIF());
+    });
+
+    app.post('/create-transaction', (req, res) => {
+        const validator = (
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/bitcoinjs-lib/src/index.ts]: done

File: examples/bitcoinjs-lib/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+            'L3BybjkmnMdXE6iNEaeZTjVMTHA4TvpYbQozc264Lto9yVDis2nv'
+        );
+        res.send(keyPair.publicKey.toString('hex'));
+    });
+
+    app.get('/get-private-key', (req, res) => {
+        const keyPair = ECPair.fromWIF(
+            'L3BybjkmnMdXE6iNEaeZTjVMTHA4TvpYbQozc264Lto9yVDis2nv'
+        );
+        keyPair.toWIF();
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/bitcoinjs-lib/src/index.ts]: done

File: examples/bitcoinjs-lib/package.json [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+        "pretest": "ts-node --transpile-only --ignore=false test/pretest.ts",
+        "test": "ts-node --transpile-only --ignore=false test/test.ts"
+    },
+    "dependencies": {
+        "@bitcoin-js/tiny-secp256k1-asmjs": "^2.2.3",
+        "azle": "0.20.1",
+        "bitcoinjs-lib": "^6.1.5",
+        "ecpair": "^2.1.0",
+        "express": "^4.18.2",
+        "i": "^0.3.7",
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/bitcoinjs-lib/package.json]: done

File: examples/bitcoinjs-lib/test/tests.ts [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+            name: '/verify-bitcoin-message',
+            test: async () => {
+                try {
+                    const response = await fetch(
+                        `${origin}/verify-bitcoin-message`,
+                        {
+                            method: 'POST'
+                        }
+                    );
+                    const responseText = await response.json();
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/bitcoinjs-lib/test/tests.ts:110]: done, fail-to-verify-bitcoin-message test has been added

File: examples/bitcore-lib/test/tests.ts [hunk 15]
```diff
@@ -0,0 +1,7 @@
+import { Test } from 'azle/test';
+
+import { getBitcoinTests } from './bitcoin';
+
+export function getTests(canisterId: string): Test[] {
+    return [...getBitcoinTests(canisterId)];
```
Comment (lastmjs) [inline examples/bitcore-lib/test/tests.ts:6]: Get rid of

File: examples/bitcore-lib/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+        'b221d9dbb083a7f33428d7c2a3c3198ae925614d70210e28716ccaa7cd4ddb79'
+    );
+    res.send(privateKey.toWIF());
+});
+
+app.post('/create-transaction', (req, res) => {
+    const privateKey = new PrivateKey(
+        'b221d9dbb083a7f33428d7c2a3c3198ae925614d70210e28716ccaa7cd4ddb79'
+    );
+    const utxo = Transaction.UnspentOutput.fromObject({
```
Comment (lastmjs) [inline examples/bitcore-lib/src/index.ts:38]: Let's figure out where we got these

File: examples/bitcoinjs-lib/src/index.ts [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+app.get('/get-private-key-wif', (req, res) => {
+    const keyPair = ECPair.fromWIF(
+        'L3BybjkmnMdXE6iNEaeZTjVMTHA4TvpYbQozc264Lto9yVDis2nv'
+    );
+
+    res.send(keyPair.toWIF());
+});
+
+app.post('/create-transaction', (req, res) => {
+    const transaction = new bitcoin.Transaction();
```
Comment (lastmjs) [inline examples/bitcoinjs-lib/src/index.ts:47]: let

File: examples/bitcoinjs-lib/src/index.ts [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+    transaction.version = 2;
+
+    const prevTxId =
+        '115e8f72f39fad874cfab0deed11a80f24f967a84079fb56ddf53ea02e308986';
+    const txHash = Buffer.from(prevTxId, 'hex').reverse();
```
Comment (lastmjs) [inline examples/bitcoinjs-lib/src/index.ts:52]: let

**Review by lastmjs — CHANGES_REQUESTED** [review]

**Review by lastmjs — APPROVED** [review]

File: examples/bitcoinjs-lib/test/pretest.ts [hunk 19]
```diff
@@ -8,6 +8,8 @@ async function pretest() {
     execSync(`dfx deploy`, {
         stdio: 'inherit'
     });
+
+    execSync(`cd ../bitcore-lib && npm install`, { stdio: 'inherit' });
```
Comment (lastmjs) [inline examples/bitcoinjs-lib/test/pretest.ts:12]: I think this deserves a comment explaining its existence

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: examples/bitcoinjs-lib/test/pretest.ts (hunk 19 above)
Comment (lastmjs) [reply examples/bitcoinjs-lib/test/pretest.ts:12]: I'm going to make an issue for this, please address it soon: https://github.com/demergent-labs/azle/issues/1745

**Review by lastmjs — APPROVED** [review]

**Pull request merged**